app/questionnaire/completeness.py

```diff
@@ -0,0 +1,188 @@
+
+    def __init__(self, schema, answer_store, completed_blocks, routing_path, metadata):
+        self.answer_store = answer_store
+        self.completed_blocks = completed_blocks
+        self.routing_path = routing_path
+        self.metadata = metadata
+        self.schema = schema
+
+    def is_section_complete(self, section):
+        return self.get_state_for_section(section) == self.COMPLETED
+
+    def is_group_complete(self, group, group_instance=None):
+        return self.get_state_for_group(group, group_instance=group_instance) == self.COMPLETED
+
+    def is_block_complete(self, location):
+        return location in self.completed_blocks
+
+    def all_sections_complete(self):
+        return all(state in [self.COMPLETED, self.SKIPPED] for state in self._get_all_section_states())
+
+    def any_section_complete(self):
+        return any(state == self.COMPLETED for state in self._get_all_section_states())
+
+    def get_state_for_section(self, section):
+        if isinstance(section, str):
+            # lookup section by section ID
+            section = self.schema.get_section(section)
+
+        group_states = [
+            self.get_state_for_group(group) for group in section['groups']
+        ]
+
+        def eval_state(state_to_compare):
+            return (state == state_to_compare for state in group_states)
+
+        section_state = self.NOT_STARTED
+
+        if group_states:
+            if all(eval_state(self.SKIPPED)):
+                section_state = self.SKIPPED
+
+            elif all(eval_state(self.INVALID)):
+                section_state = self.INVALID
+
+            elif any(eval_state(self.STARTED)):
+                section_state = self.STARTED
+
+            elif all(state in self.COMPLETED_STATES for state in group_states):
+                section_state = self.COMPLETED
+
+        return section_state
+
+    def get_state_for_group(self, group, group_instance=None):
+        if isinstance(group, str):
+            # lookup group by group ID
+            group = self.schema.get_group(group)
+
+        if (QuestionnaireSchema.is_confirmation_group(group) or
+                QuestionnaireSchema.is_summary_group(group)):
+            # summary/confirmations are special cases as we don't want to
+            # render until the whole survey is complete. They're also never
+            # show as complete
+            return self.NOT_STARTED if self.all_sections_complete() else self.SKIPPED
+
+        if self._should_skip(group):
+            return self.SKIPPED
+
+        block_states = [
+            state for location, state in
+            self._get_block_states_for_group(group, group_instance)
+        ]
+
+        def eval_state(state_to_compare):
+            return (state == state_to_compare for state in block_states)
+
+        group_state = self.NOT_STARTED
+
+        if all(eval_state(self.SKIPPED)):
+            group_state = self.SKIPPED
+
+        elif not self.routing_path:
```

Comment:
A lot of duplication here. Not very familiar with methods within methods. Can these be simplified into the one method?

Reply:
@nealedj, are you sure that we can't reduce the duplication around eval_state. That looks the same for group and block.

Reply:
It would save a good few lines of code 😉

Reply:
Thats fine! I thought originally we might be able to, was worth asking the question